src/devices/Common/Iot/Device/Common/I2cRegisterAccess.cs

```diff
@@ -0,0 +1,120 @@
+    ///     Register2 = 0x01,
+    ///     Register3 = 0x03
+    /// }
+    /// </code>
+    /// </example>
+    /// </summary>
+    public sealed class I2cRegisterAccess<TRegisterMap> : IDisposable
+        where TRegisterMap : Enum
+    {
+        private I2cDevice _device;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="I2cRegisterAccess{TRegisterMap}"/> class.
+        /// </summary>
+        /// <param name="device">I2C device to access.</param>
+        public I2cRegisterAccess(I2cDevice device)
+        {
+            _device = device;
+        }
+
+        /// <summary>
+        /// Reads an 8-bit integer from the provided register address.
+        /// </summary>
+        public byte ReadSingleRegister(TRegisterMap register)
+        {
+            Span<byte> readBuffer = new byte[1];
+            _device.WriteRead(ToReadOnlySpan(register), readBuffer);
+            if (readBuffer.IsEmpty)
+            {
+                return 0;
+            }
+
+            return MemoryMarshal.Read<byte>(readBuffer);
+        }
+
+        /// <summary>
+        /// Reads a 16-bit integer from the provided register address.
+        /// </summary>
+        public ushort ReadDoubleRegister(TRegisterMap register)
+        {
+            Span<byte> readBuffer = new byte[2];
+            _device.WriteRead(ToReadOnlySpan(register), readBuffer);
+            if (readBuffer.IsEmpty)
+            {
+                return 0;
+            }
+
+            return MemoryMarshal.Read<ushort>(readBuffer);
+        }
+
+        /// <summary>
+        /// Reads a 32-bit integer from the provided register address.
+        /// </summary>
+        public uint ReadQuadRegister(TRegisterMap register)
+        {
+            Span<byte> readBuffer = new byte[4];
+            _device.WriteRead(ToReadOnlySpan(register), readBuffer);
+            if (readBuffer.IsEmpty)
+            {
+                return 0;
+            }
+
+            return MemoryMarshal.Read<uint>(readBuffer);
+        }
+
+        /// <summary>
+        /// Writes an 8-bit integer to the provided register address.
+        /// </summary>
+        public void WriteSingleRegister(TRegisterMap register, byte data)
+        {
+            _device.Write(new[] { (byte)(object)register, data });
+        }
+
+        /// <summary>
+        /// Writes a 16-bit integer to the provided register address.
+        /// </summary>
+        public void WriteDoubleRegister(TRegisterMap register, ushort data)
+        {
+            byte lower = (byte)(data & 0xff);
+            byte upper = (byte)(data >> 8);
+            _device.Write(new ReadOnlySpan<byte>(new[]
+            {
+               (byte)(object)register,
+               lower,
+               upper
+            }));
+        }
+
+        private static ReadOnlySpan<byte> ToReadOnlySpan(TRegisterMap registerValue)
+        {
+            return new ReadOnlySpan<byte>(new[] { (byte)(object)registerValue });
```

Comment:
this is missing the purpose of Span. You use Span in order to avoid allocation, in here you neither have the benefit of the array nor span

Reply:
I've completely rewritten I2cRegisterAccess and made it into the two classes I2cRegisterAccess and I2cRegisterAccess<T>.
I also moved them back into the QwiicButton solution. If you like them, we can later move them into the CommonHelpers solution.